[improve][doc] Improve golang client doc

labuladong opened the pull request:
<!--
### Contribution Checklist
  
  - PR title format should be *[type][component] summary*. For details, see *[Guideline - Pulsar PR Naming Convention](https://docs.google.com/document/d/1d8Pw6ZbWk-_pCKdOmdvx9rnhPiyuxwq60_TrD68d7BA/edit#heading=h.trs9rsex3xom)*. 

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.

**(The sections below can be removed for hotfixes of typos)**
-->



Fixes #17397

### Motivation

Fix some errors in the Golang doc and improve code demos.

### Modifications

Update some out-of-date API interfaces and make all code snippets can run.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `doc-not-needed` 
(Please explain why)
  
- [x] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

tisonkun commented:
Comment for a cross reference to https://github.com/apache/pulsar/issues/17302.

labuladong commented:
cc @Anonymitaet @momo-jun

Anonymitaet commented on site2/docs/client-libraries-go.md:
```diff
@@ -846,43 +868,43 @@ ctx := context.Background()
-	assert.NoError(t, err)
-	assert.NotNil(t, msgID)
-	msgIDs[i] = msgID
+    msgID, _ := producer.Send(ctx, &pulsar.ProducerMessage{
```
@nodece 
could you please review this PR from a technical perspective? Thank you!

labuladong replied:
Seems that the original code snippet is copied from the test cases, which is unnecessary in the documentation.

nodece replied:
Hi @Anonymitaet, I suggest we remove parameter details, and then use https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar instead, when the go-client updates, we need to update our documentation again, what do you think of this.

labuladong replied:
I suggest reserving them because other clients' documentation also has parameter details.

RobertIndie replied:
+1 for removing them. It's hard to maintain the parameters detail in two places: the go doc and the pulsar doc website. Or we can provide some commonly used parameters in this doc and let's user jump to go doc for more detail.
In fact, the java client doc also encourages the user to jump to the java doc to get more details. But leave the loadConf configuration which is generated from the code in the doc.

Anonymitaet replied:
@nodece I'm on your side. Referencing the source material reduces maintenance cost

@labuladong FYI: to reduce manual work, we (@SignorMercurio @urfreespace) are trying to generate client configurations docs from code automatically (https://github.com/apache/pulsar/pull/17198), which means we'll remove the [configuration instructions - table](https://pulsar.apache.org/docs/next/client-libraries-java#configure-producer) from client doc.

labuladong replied:
Got it. I will remove the parameter details and link to https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar

nodece commented on site2/docs/client-libraries-go.md:
````diff
@@ -538,37 +556,39 @@ func main() {
+#### Receive message with timeout
 
 ```go
 client, err := NewClient(pulsar.ClientOptions{
````
```suggestion
client, err := pulsar.NewClient(pulsar.ClientOptions{
```

nodece commented on site2/docs/client-libraries-go.md:
````diff
@@ -608,16 +629,12 @@ type testJSON struct {
-```
 
-```go
 client, err := NewClient(pulsar.ClientOptions{
````
```suggestion
client, err := pulsar.NewClient(pulsar.ClientOptions{
```

nodece replied:
`consumerJS := NewJSONSchema(exampleSchemaDef, nil)` -> `consumerJS := pulsar.NewJSONSchema(exampleSchemaDef, nil) `

nodece reviewed (changes requested):
Overall this PR looks good to me, I noticed we added some configurable parameters on this page.

I suggest link to https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar

labuladong commented:
Thanks for your suggestion. Links added. @nodece

RobertIndie commented on site2/docs/client-libraries-go.md:
```diff
@@ -442,85 +433,114 @@ if err := consumer.Unsubscribe(); err != nil {
-client, err := pulsar.NewClient(pulsar.ClientOptions{
-    URL: "pulsar://localhost:6650",
-})
+client, err := pulsar.NewClient(pulsar.ClientOptions{URL: "pulsar://localhost:6650"})
+if err != nil {
+    log.Fatal(err)
+}
 
 defer client.Close()
 
-p, err := client.CreateProducer(pulsar.ProducerOptions{
-	Topic:           topicInRegex,
-	DisableBatching: true,
+// subscribe a specific partition of a topic
+consumer, err := client.Subscribe(pulsar.ConsumerOptions{
+    // pulsar partition is a special topic has the suffix '-partition-xx'
+    Topic:            "my-partitioned-topic-partition-2",
+    SubscriptionName: "my-sub",
+    Type:             pulsar.Shared,
 })
```
Why do we add this example? I think subscribing to a specific partition is not a good practice.

labuladong replied:
This example splits from here: 
https://pulsar.apache.org/docs/next/client-libraries-go#how-to-use-message-router-in-producer

RobertIndie replied:
The original example is to introduce how to use the message router. It's from the producer part and is just used to test the custom message router.

labuladong replied:
You are right, I'll remove this example.

labuladong commented:
I've removed the partition subscription example [1aeac82](https://github.com/apache/pulsar/pull/17455/commits/1aeac82a64cf0b1fb04a07ec82511cc4800d376c) and replaced all configuration with link [6298411](https://github.com/apache/pulsar/pull/17455/commits/629841176313a1f0a4ee790bea24027565fd68d6).

PTAL @nodece @Anonymitaet @RobertIndie

Anonymitaet commented on site2/docs/client-libraries-go.md:
````diff
@@ -442,85 +420,75 @@ if err := consumer.Unsubscribe(); err != nil {
 ```
 
-#### How to use multi topics Consumer
+#### Create multi-topic Consumer
````
```suggestion
#### Create multi-topic consumer
```

Anonymitaet commented on site2/docs/client-libraries-go.md:
````diff
@@ -585,64 +555,66 @@ defer cancel()
+fmt.Println(msg.Payload())
 ```
 
 #### How to use schema in consumer
````
```suggestion
#### Use schema in consumer
```

Anonymitaet commented on site2/docs/client-libraries-go.md:
```diff
@@ -832,12 +774,12 @@ reader, err := client.CreateReader(pulsar.ReaderOptions{
 #### How to use reader to read specific message
```
```suggestion
#### Use reader to read specific message
```

tisonkun commented on site2/docs/client-libraries-go.md:
```diff
@@ -424,6 +401,7 @@ if err != nil {
 defer consumer.Close()
 
 for i := 0; i < 10; i++ {
+    // may block here
 	msg, err := consumer.Receive(context.Background())
 	if err != nil {
 		log.Fatal(err)
```
@labuladong generally Golang use tab for indent. In this patch we mix tabs and spaces, you may keep them consistent.

tisonkun replied:
Not really. [These lines](https://github.com/apache/pulsar/blob/a053323e23998d9b7e746933cbce43590ad956b9/site2/docs/client-libraries-go.md?plain=1#L71-L76) are still using tabs. I suggest you replace all leading space in Golang code blocks with tabs.

labuladong replied:
Code in markdown block will treat `\t` as 4 spaces. But you are right, I can use regex to replace them.

tisonkun replied:
Yes. I'm OK to use spaces here since Go formatter should be able to format it. My suggestion focuses on consistently - that is, we don't use both.

Anonymitaet commented:
Thanks for your contribution! Have you previewed your changes to ensure everything goes as expected? 

If not yet, please [preview your changes locally](https://github.com/apache/pulsar-site/tree/main/site2/website-next#local-preview) and attach the screenshots in the PR description to let others know. In this way, you can get your PR merged more quickly. Thanks!

labuladong commented:
Screenshots of this PR:

![FireShot Capture 003 - Pulsar Go client - Apache Pulsar - localhost](https://user-images.githubusercontent.com/37220920/188593890-227d7397-086e-4f67-b754-07c18739ef67.jpg)

PTAL @Anonymitaet

momo-jun commented on site2/docs/client-libraries-go.md:
```diff
@@ -716,63 +688,33 @@ scrape_configs:
-`Close()` | Closes the reader, disabling its ability to receive messages from the broker | `error`
-`Seek(MessageID)` | Reset the subscription associated with this reader to a specific message ID | `error`
-`SeekByTime(time time.Time)` | Reset the subscription associated with this reader to a specific message publish time | `error`
+All available method of `Reader` interface are [here](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#Reader).
```
```suggestion
All available methods of the `Reader` interface are [here](https://pkg.go.dev/github.com/apache/pulsar-client-go/pulsar#Reader).
```

labuladong commented:
ping @Anonymitaet
